### ISSUE TITLE
Performance degradation, after performance tests with big (4K) payloads.

### ISSUE THREAD

**Issue by AlexanderKlishin** [tarantool/tarantool#4736]
Tarantool version:
tarantool-2.1.3.6-1.el7.x86_64

OS version:
Red Hat Enterprise Linux Server release 7.7 (Maipo)

Bug description:
Performance degradation, after performance tests with big (4K) payloads.

Steps to reproduce:
1. create space + primary index
local space = box.schema.space.create('ipcan', {
    if_not_exists = true,
     engine = "memtx"
})
space:create_index('pk', {
    type = "tree",
    parts = { 1, "string", 2, "integer", 3, "integer" },
    if_not_exists = true
})

2. create api function
function api_upsert(k1, k2, k3, data)
    box.space.ipcan:upsert({ k1, k2, k3, data },  {{ '=', 4, data } })
  end

3. call api_upsert("1", 2, 3, "data length 1K")
Run load for 10 sec
Result performace - 60K/sec

4. call api_upsert("1", 2, 3, "data length 4K")
Run load for 10 sec
Result performance - 6K

5. call api_upsert("1", 2, 3, "data length 1K")
At this point - expect 60K/sec, but got only 6K/sec

P.S.
Tests on version 1.10 show no performance degradation on all steps.

**Comment by AlexanderKlishin** [comment]
Tested version: tarantool-2.2.2.17-1.el7.x86_64.
The results are the same: 10 times performance degradation after performance tests with 4K payload.

The only working version for my workload is 1.10.

But i need "on_schema_init", to start some triggers before replication process is started.
"on_schema_init" implemented in 2.X versions, so this is a problem for me - to use 1.10 in production.

**Comment by Totktonada** [comment]
TL;DR: unable to reproduce. More information is necessary to reproduce the problem.

Note: I made upserts against one tuple as shown above. So almost all upserts were made against the existing tuple (they become updates). If it is assumed that keys should be different, then I should know at least a percent of upserts that is made against existing tuples and ones against non-existing tuples.

## My steps

Before build:

```sh
$ mkdir 2.1 && cd 2.1
$ git clone --recursive --branch 2.1 git@github.com:tarantool/tarantool.git
$ cd tarantool
$ git clean -xfd; git submodule foreach --recursive 'git clean -xfd'
$ cd ../..
```

How to build:

```sh
$ cmake . -DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_BACKTRACE=ON -DENABLE_DIST=ON -DENABLE_BUNDLED_LIBCURL=OFF && make -j
```

Copy the following files to the current directory:

box.lua:

```lua
#!/usr/bin/env tarantool

box.cfg({
    listen = 3301,
    background = true,
    log = 'box.log',
    pid_file = 'box.pid',
})
box.schema.user.grant('guest', 'read,write,execute', 'universe')

box.schema.space.create('s', {
    if_not_exists = true,
    engine = 'memtx',
})
box.space.s:create_index('pk', {
    type = 'tree',
    parts = {1, 'string', 2, 'integer', 3, 'integer'},
    if_not_exists = true,
})

function api_upsert(k1, k2, k3, data)
    box.space.s:upsert({k1, k2, k3, data}, {{'=', 4, data}})
end
```

workload.lua:

```lua
#!/usr/bin/env tarantool

local net_box = require('net.box')
local clock = require('clock')
local fiber = require('fiber')

local BATCH_COUNT = 1000

local data_size = tonumber(arg[1])
local data = string.rep('x', data_size)

local connection = net_box.connect('localhost:3301')

-- {{{ Parallel requests

local BATCH_SIZE = 1000

local function feed_batch()
    local futures = {}
    for i = 1, BATCH_SIZE do
        local f = connection:call('api_upsert', {'1', 2, 3, data}, {is_async = true})
        table.insert(futures, f)
    end
    for _, f in ipairs(futures) do
        f:wait_result()
    end
end

-- }}}

-- {{{ Non-parallel requests

--[[
local BATCH_SIZE = 50

local function feed_batch()
    for i = 1, BATCH_SIZE do
        connection:call('api_upsert', {'1', 2, 3, data})
    end
end
]]--

-- }}}

local function print_rps(prefix, start_time, end_time, request_count)
    local duration = end_time - start_time
    local rps = request_count / duration
    print(prefix .. ('%f'):format(rps))
end

local prefix = ('Batch RPS (%d): '):format(data_size)
local start_time = clock.monotonic()
for i = 1, BATCH_COUNT do
    local batch_start_time = clock.monotonic()
    feed_batch()
    local batch_end_time = clock.monotonic()
    --print_rps('Batch RPS: ', batch_start_time, batch_end_time, BATCH_SIZE)
end
local end_time = clock.monotonic()
local prefix = ('Overall RPS (%d): '):format(data_size)
print_rps(prefix, start_time, end_time, BATCH_SIZE * BATCH_COUNT)

os.exit()
```

stat.sh:

```sh
#!/bin/sh

awk '
    BEGIN {
        min = "NaN"
        max = "NaN"
        cnt = 0
        var = "NaN"
        sum = "NaN"
        avg = "NaN"
    }
    {
        min = (NR==1 || $1<min ? $1 : min)
        max = (NR==1 || $1>max ? $1 : max)
        cnt += 1
        sum += $1
    }
    END {
        avg = sum / cnt
        var = max - min
        var_rel = 100 * var / avg
        print "cnt:", cnt
        print "min:", min
        print "max:", max
        print "avg:", avg
        print "var:", var, "("var_rel"%)"
        print ""
    }
';
```

Makefile:

```make
default: run

#T="./1.10/tarantool/src/tarantool"
T="./2.1/tarantool/src/tarantool"
#T="./master/tarantool/src/tarantool"

.PNOHY: run
run:
	rm -f *.{snap,xlog} box.log box.pid res.txt
	$T --version | head -n 2
	$T box.lua
	sleep 1
	$T workload.lua 1000 | tee -a res.txt
	$T workload.lua 4000 | tee -a res.txt
	$T workload.lua 1000 | tee -a res.txt
	kill $$(cat box.pid)
	cat res.txt | awk '{ print $$4 }' | sh stat.sh
```

Run `make` to start tarantool, feed workload and show statistics.

## 1.10 results

Tarantool 1.10.5-6-g764eec8b5
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 182829.726505
Overall RPS (4000): 113213.213709
Overall RPS (1000): 190701.264427
cnt: 3
min: 113213.213709
max: 190701.264427
avg: 162248
var: 77488.1 (47.759%)

Tarantool 1.10.5-6-g764eec8b5
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 181864.353342
Overall RPS (4000): 110272.006995
Overall RPS (1000): 182460.919316
cnt: 3
min: 110272.006995
max: 182460.919316
avg: 158199
var: 72188.9 (45.6317%)

Tarantool 1.10.5-6-g764eec8b5
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 181529.343153
Overall RPS (4000): 115845.260948
Overall RPS (1000): 197184.732798
cnt: 3
min: 115845.260948
max: 197184.732798
avg: 164853
var: 81339.5 (49.3406%)

## 2.1 results

Tarantool 2.1.3-6-g91e2a9638
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 173142.303705
Overall RPS (4000): 115572.727931
Overall RPS (1000): 172830.814161
cnt: 3
min: 115572.727931
max: 173142.303705
avg: 153849
var: 57569.6 (37.4196%)

Tarantool 2.1.3-6-g91e2a9638
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 175766.817835
Overall RPS (4000): 115115.038478
Overall RPS (1000): 168731.546698
cnt: 3
min: 115115.038478
max: 175766.817835
avg: 153204
var: 60651.8 (39.5888%)

Tarantool 2.1.3-6-g91e2a9638
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 170710.995398
Overall RPS (4000): 117112.386480
Overall RPS (1000): 163927.056311
cnt: 3
min: 117112.386480
max: 170710.995398
avg: 150583
var: 53598.6 (35.594%)

## master results

Tarantool 2.4.0-16-gcdf502c66
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 160980.816268
Overall RPS (4000): 106476.499264
Overall RPS (1000): 159655.794190
cnt: 3
min: 106476.499264
max: 160980.816268
avg: 142371
var: 54504.3 (38.2833%)

Tarantool 2.4.0-16-gcdf502c66
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 159771.146953
Overall RPS (4000): 107861.893967
Overall RPS (1000): 143784.370798
cnt: 3
min: 107861.893967
max: 159771.146953
avg: 137139
var: 51909.3 (37.8515%)

Tarantool 2.4.0-16-gcdf502c66
Target: Linux-x86_64-RelWithDebInfo
Overall RPS (1000): 160638.644711
Overall RPS (4000): 109739.376152
Overall RPS (1000): 158549.773457
cnt: 3
min: 109739.376152
max: 160638.644711
avg: 142976
var: 50899.3 (35.5999%)

## Interpretation

Compared using resuls of 1000 bytes runs:

* ~8.2% degradation between 1.10 and 2.1.
* ~8.0% degradation between 2.1 and master.
* ~15.5% degradation between 1.10 and master.

However the effect of several times degradation of small requests after large ones is not seen.

`perf top -a -g` shows that most time is spent in Lua, but upsert itself consumes very small fraction of the execution time.

## One more experiment

I run also several non-parallel runs on 2.1 (see commented code in workload.lua), but it does not show the reported effect too.

**Comment by rybakit** [comment]
I've noticed a typo:

```diff
-.PNOHY: run
+.PHONY: run
```

**Comment by AlexanderKlishin** [comment]
Version for tests: tarantool-2.1.3.6-1.el7.x86_64
(My new virtual host is relatively slow - only 40K/sec)

The only difference between tests - is payload size.
Try to run workload.lua with **4k (4096)** payload size.

I have changed local BATCH_COUNT = 100
```
$ tarantool workload.lua 1000
Overall RPS (1000): 45796.865030
$ tarantool workload.lua 4000
Overall RPS (4000): 39137.994332
$ tarantool workload.lua 4000
Overall RPS (4000): 35803.389764
$ tarantool workload.lua 4000
Overall RPS (4000): 37865.379561
```

everything is ok, until this point

```
$ tarantool workload.lua 4096
Overall RPS (4096): 10761.374278
$ tarantool workload.lua 4096
Overall RPS (4096): 4961.889773
$ tarantool workload.lua 4096
Overall RPS (4096): 3218.230914
$ tarantool workload.lua 1000
Overall RPS (1000): 2984.985532
$ tarantool workload.lua 1000
Overall RPS (1000): 3020.839328
$ tarantool workload.lua 1000
Overall RPS (1000): 3078.025442
```

10 times degradation.

```
perf top -g -p PID:
-   97.52%     0.09%  tarantool  tarantool           [.] lj_BC_FUNCC                                                                                                                 
   - 97.44% lj_BC_FUNCC                                                                                                                                                              
      - 96.89% lbox_upsert                                                                                                                                                           
         - 94.93% box_upsert                                                                                                                                                         
            - 94.91% box_process_rw                                                                                                                                                  
                 93.65% fiber_gc                                                                                                                                                     
               + 0.66% space_execute_dml                                                                                                                                             
                 0.58% txn_commit_stmt                                                                                                                                               
         + 1.95% lbox_encode_tuple_on_gc                                                                                                                                             
-   96.89%     0.01%  tarantool  tarantool           [.] lbox_upsert                                                                                                                 
   - 96.88% lbox_upsert                                                                                                                                                              
      - 94.93% box_upsert                                                                                                                                                            
         - 94.91% box_process_rw                                                                                                                                                     
              93.65% fiber_gc                                                                                                                                                        
            + 0.66% space_execute_dml                                                                                                                                                
              0.58% txn_commit_stmt                                                                                                                                                  
      + 1.95% lbox_encode_tuple_on_gc                                                                                                                                                
-   94.93%     0.01%  tarantool  tarantool           [.] box_upsert                                                                                                                  
   + 94.92% box_upsert                                                                                                                                                               
-   94.91%     0.00%  tarantool  tarantool           [.] box_process_rw                                                                                                              
   - 94.90% box_process_rw                                                                                                                                                           
        93.65% fiber_gc                                                                                                                                                              
      + 0.66% space_execute_dml                                                                                                                                                      
        0.58% txn_commit_stmt                                                                                                                                                        
+   93.65%    93.65%  tarantool  tarantool           [.] fiber_gc                                                                                                                    
+   21.85%     0.00%  tarantool  [unknown]           [.] 0x00000000417513b8                                                                                                          
+   11.05%     0.00%  tarantool  [unknown]           [.] 0xffffffff417513b8                                                                                                          
+    1.95%     0.02%  tarantool  tarantool           [.] lbox_encode_tuple_on_gc                                                                                                     
+    1.86%     0.01%  tarantool  tarantool           [.] luamp_encode_tuple                                                                                                          
+    1.85%     0.00%  tarantool  tarantool           [.] luamp_encode                                                                                                                
```

**Comment by Totktonada** [comment]
Reproduced now, thanks! Changed data sizes to 1024 and 4096.

Bisect log:

```
2.1.0-0-g2c6929204    good
2.1.1-312-g305dbcf68  good
2.1.2-16-g847061daa   good
2.1.2-41-gc69c0a639   good
2.1.2-54-g8f954a745   good
2.1.2-60-g5ac249bb2   good
2.1.2-63-g1df5c6228   good
2.1.2-65-geca3411a4   good
2.1.2-66-g3b19db722   good
2.1.2-67-g586b701a8   bad
2.1.2-118-g661f7171f  bad
2.1.3-6-g91e2a9638    bad
```

```
$ git show 586b701a88a1e1d1c98db01efe04ad93484785b3 | cat
commit 586b701a88a1e1d1c98db01efe04ad93484785b3
Author: Georgy Kirichenko <georgy@tarantool.org>
Date:   Sat May 18 23:25:08 2019 +0300

    Update small library
    
    (cherry picked from commit e4fc22d00d1a9088631fe8eae0d08157834d94d8)

diff --git a/src/lib/small b/src/lib/small
index cdf7d4a8e..67d7ab44a 160000
--- a/src/lib/small
+++ b/src/lib/small
@@ -1 +1 @@
-Subproject commit cdf7d4a8e24ae465298d0ddacaf1aaed1085281e
+Subproject commit 67d7ab44ab09df3356929e3692a03321b31f3ebb
```

It gives the following commits:

```
$ (cd src/lib/small && git log --oneline cdf7d4a8e24ae465298d0ddacaf1aaed1085281e..67d7ab44ab09df3356929e3692a03321b31f3ebb | cat)
67d7ab4 Free all slabs on region reset
76be096 small: fix false-positive 'out of array bound' error
45710a9 small: fix a typo in a function name
d97f2a0 small: introduce static allocator
```

Checked out 586b701a88a1e1d1c98db01efe04ad93484785b3 and reverted the last src/lib/small commit (67d7ab4 Free all slabs on region reset). After that the performance got back.

So, the cause is https://github.com/tarantool/small/commit/67d7ab44ab09df3356929e3692a03321b31f3ebb

**Comment by AlexanderKlishin** [comment]
When this fix will be available in 2.1 branch ? :)
